compiler/rustc_middle/src/mir/interpret/mod.rs

```diff
@@ -259,7 +258,7 @@ pub enum GlobalAlloc<'tcx> {
     /// The alloc ID is used as a function pointer.
     Function { instance: Instance<'tcx> },
     /// This alloc ID points to a symbolic (not-reified) vtable.
-    VTable(Ty<'tcx>, Option<ty::PolyExistentialTraitRef<'tcx>>),
```

Comment:
Codegen fully identifies a vtable using just the `Option<ty::PolyExistentialTraitRef<'tcx>>`:

https://github.com/rust-lang/rust/blob/bdacdfe95f17188f5b047e0330e3e47f993a2016/compiler/rustc_codegen_ssa/src/meth.rs#L118

So how can it be that Miri would require more than this? Is it just that currently it happens to be the case that vtables only depend on the principal (i.e., never depend on projections or auto traits) but we want to reserve the right to change this later?

Reply:
Codegen fully identifies a vtable using a `Option<PolyExistentialTraitRef>` *and* a self `Ty`.

Miri doesn't require the full set of predicates, either, technically, but computing the set of projections that correspond to a vtable requires also using the self type to compute what the projections would be. We could also pass that in to the validity, since it's stored in the allocation, but that would be pretty expensive to test a bunch of projection goals every time we want to validate vtable equality.

Reply:
For the record, that corresponds roughly to my alternative here: 

https://github.com/rust-lang/rust/pull/130727#issuecomment-2367075142